core/app/models/spree/weighted_adjustments_handler.rb

```diff
@@ -0,0 +1,47 @@
+module Spree
+  class WeightedAdjustmentsHandler
+    attr_reader :line_item, :order, :preferred_amount
+
+    def initialize(line_item, preferred_amount)
+      @line_item = line_item
+      @order = line_item.order
+      @preferred_amount = preferred_amount
+    end
+
+    # Returns the weighted adjustment for the initialized line item
+    #
+    # @return [Integer]
+    def amount
+      weighted_amounts[@line_item.id].to_f
+    end
+
+    private
+
+    # Returns a hash of line item IDs and their corresponding weighted
+    # adjustments
+    #
+    # @private
+    # @return [Hash<Integer, Integer>]
+    def weighted_amounts
+      remaining_amount = @preferred_amount
+
+      @order.line_items.each_with_index.map do |line_item, i|
+        if i == @order.line_items.count - 1
+          # If this is the last line item on the order we want to use the
+          # remaining preferred amount to ensure our total adjustment is what
+          # has been set as the preferred amount.
+          [line_item.id, remaining_amount]
+        else
+          # Calculate the weighted amount by getting this line item's share of
+          # the order's total and multiplying it with the preferred amount.
+          weighted_amount = ((line_item.amount / @order.item_total) * preferred_amount).round(2)
```

Comment:
I'm happy to leave this for now. Would prefer to avoid tacking too many things onto this PR.